Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,730 @@
+  ref->udp_socket.setBlocking(false);
+  if (ref->udp_socket.bind(port) != sf::Socket::Done)
+  {
+    ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+    return;
+  }
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::HandleUDPFrame(Common::EthernetHeader* hwdata,
+                                                       Common::IPv4Header* ipdata,
+                                                       Common::UDPHeader* udpdata, u8* data)
+{
+  sf::IpAddress target;
+
+  if (*(u32*)ipdata->destination_addr == 0)
+    *(u32*)ipdata->destination_addr = m_router_ip;
+  // dns request
+  StackRef* ref = GetAvaibleSlot(udpdata->source_port);
+  if (ref->ip == 0)
+  {
+    ref->ip = *(u32*)ipdata->destination_addr;  // change for ip
+    ref->local = udpdata->source_port;
+    ref->remote = udpdata->destination_port;
+    ref->type = 17;
+    ref->bba_mac = *(Common::MACAddress*)&m_eth_ref->mBbaMem[BBA_NAFR_PAR0];
+    ref->my_mac = fake_mac;
+    ref->from.sin_addr.s_addr = *(u32*)&ipdata->destination_addr;
+    ref->from.sin_port = udpdata->destination_port;
+    ref->to.sin_addr.s_addr = *(u32*)&ipdata->source_addr;
+    ref->to.sin_port = udpdata->source_port;
+    ref->udp_socket.setBlocking(false);
+    if (ref->udp_socket.bind(htons(udpdata->source_port)) != sf::Socket::Done &&
+        ref->udp_socket.bind(sf::Socket::AnyPort) != sf::Socket::Done)
+    {
+      ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+      return;
+    }
+  }
+  if (Common::swap16(udpdata->destination_port) == 1900)
+  {
+    InitUDPPort(26512);                                 // MK DD and 1080
+    InitUDPPort(26502);                                 // Air Ride
+    if (*(u32*)ipdata->destination_addr == 0xFAFFFFEF)  // force real broadcast
+      *(u32*)ipdata->destination_addr = 0xFFFFFFFF;     // Multi cast cannot be read
+    if (udpdata->length > 150)
+    {
+      // Quick hack to unlock the connection, throw it back at him
+      Common::EthernetHeader* hwpart = (Common::EthernetHeader*)m_in_frame.get();
+      Common::IPv4Header* ippart = (Common::IPv4Header*)&m_in_frame[14];
+      std::memcpy(m_in_frame.get(), hwdata, htons(ipdata->total_len) + 14);
+      hwpart->destination = hwdata->source;
+      hwpart->source = hwdata->destination;
+      *(u32*)ippart->destination_addr = *(u32*)ipdata->source_addr;
+      *(u32*)ippart->source_addr = *(u32*)ipdata->destination_addr;
+      WriteToQueue(m_in_frame.get(), htons(ipdata->total_len) + 14);
+    }
+  }
+  if (Common::swap16(udpdata->destination_port) == 53)
+  {
+    target = sf::IpAddress(m_dns_ip.c_str());  // dns server ip
+  }
+  else
+  {
+    target = sf::IpAddress(Common::swap32(*(u32*)ipdata->destination_addr));
+  }
+  ref->udp_socket.send(data, Common::swap16(udpdata->length) - 8, target,
+                       Common::swap16(udpdata->destination_port));
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::SendFrame(const u8* frame, u32 size)
+{
+  Common::EthernetHeader* hwdata;
```

Comment:
Are these forward declarations necessary? If not, the declaration and initialisation should be done closer to where they'll be used. Braces can be used with case statement if needed.